hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/BlobRenameHandler.java

```diff
@@ -0,0 +1,452 @@
+        LOG.info(
+            "Rename src: {} dst: {} failed as qualifiedDst already exists",
+            src, dst);
+        throw new AbfsRestOperationException(
+            HttpURLConnection.HTTP_CONFLICT,
+            AzureServiceErrorCode.PATH_ALREADY_EXISTS.getErrorCode(), null,
+            null);
+      }
+    }
+
+    if (!dst.isRoot() && nestedDstParent != null && !nestedDstParent.isRoot()
+        && (
+        !pathInformation.getIsDirectory() || !dst.getName()
+            .equals(src.getName()))) {
+      PathInformation nestedDstInfo = getPathInformation(
+          nestedDstParent,
+          tracingContext);
+      if (!nestedDstInfo.getPathExists() || !nestedDstInfo.getIsDirectory()) {
+        throw new AbfsRestOperationException(
+            HttpURLConnection.HTTP_NOT_FOUND,
+            RENAME_DESTINATION_PARENT_PATH_NOT_FOUND.getErrorCode(), null,
+            new Exception(
+                RENAME_DESTINATION_PARENT_PATH_NOT_FOUND.getErrorCode()));
+      }
+    }
+
+    return true;
+  }
+
+  @Override
+  boolean takeAction(final Path path) throws IOException {
+    return renameInternal(path,
+        createDestinationPathForBlobPartOfRenameSrcDir(dst, path, src));
+  }
+
+  private boolean renameInternal(final Path path,
+      final Path destinationPathForBlobPartOfRenameSrcDir) throws IOException {
+    final String leaseId;
+    AbfsLease abfsLease = null;
+    if (isAtomicRename) {
+      /*
+       * To maintain atomicity of rename of the path, lease is taken on the path.
+       */
+      if (path.equals(src)) {
+        abfsLease = srcAbfsLease;
+        leaseId = srcLeaseId;
+      } else {
+        abfsLease = takeLease(path, null);
+        leaseId = abfsLease.getLeaseID();
+      }
+    } else {
+      leaseId = null;
+    }
+    copyPath(path, destinationPathForBlobPartOfRenameSrcDir, leaseId);
+    abfsClient.deleteBlobPath(path, leaseId, tracingContext);
+    if (abfsLease != null) {
+      abfsLease.cancelTimer();
+    }
+    operatedBlobCount.incrementAndGet();
+    return true;
+  }
+
+  private void copyPath(final Path src, final Path dst, final String leaseId)
+      throws IOException {
+    String copyId;
+    try {
+      AbfsRestOperation copyPathOp = abfsClient.copyBlob(src, dst, leaseId,
+          tracingContext);
+      final String progress = copyPathOp.getResult()
+          .getResponseHeader(X_MS_COPY_STATUS);
+      if (COPY_STATUS_SUCCESS.equalsIgnoreCase(progress)) {
+        return;
+      }
+      copyId = copyPathOp.getResult()
+          .getResponseHeader(X_MS_COPY_ID);
+    } catch (AbfsRestOperationException ex) {
+      if (ex.getStatusCode() == HttpURLConnection.HTTP_CONFLICT) {
+        AbfsRestOperation dstPathStatus = abfsClient.getPathStatus(
+            dst.toUri().getPath(),
+            false, tracingContext, null);
+        if (dstPathStatus.getResult() != null && ((ROOT_PATH
+            + abfsClient.getFileSystem() + src.toUri()
```

Comment:
Can getRelativePath or some other existing method be called instead of building the path here ?

Reply:
Tried other different util methods but couldn't get the required heuristic.